lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
       continue;
     }
 
-    auto &component = interfaceMap[{defName.getValue(), name.getValue()}];
-    component.found = true;
-
-    switch (component.tpe) {
-    case ElementInfo::Vector:
-      component.width = type.getBitWidthOrSentinel();
-      component.depth++;
-      component.locations.push_back(op->getLoc());
-      break;
-    case ElementInfo::Ground:
-      component.width = type.getBitWidthOrSentinel();
-      component.locations.push_back(op->getLoc());
-      break;
-    case ElementInfo::Bundle:
-    case ElementInfo::String:
-    case ElementInfo::Boolean:
-    case ElementInfo::Integer:
-    case ElementInfo::Double:
-      break;
-    case ElementInfo::Error:
-      llvm_unreachable("Shouldn't be here");
-      break;
-    }
-    annotations.removeAnnotation(anno);
+    builder.create<sv::InterfaceSignalOp>(
+        uloc, name.getValue(), std::get<Type>(elementType.getValue()));
   }
-}
 
-/// Combined logic to handle Wires, Registers, and Nodes because these all use
-/// the same approach.
-void GrandCentralVisitor::handleRef(mlir::Operation *op) {
-  auto annotations = AnnotationSet(op);
-  handleRefLike(op, annotations, op->getResult(0).getType().cast<FIRRTLType>());
-  annotations.applyToOperation(op);
+  return iface;
 }
 
-/// Remove Grand Central Annotations from ports of modules or external modules.
-/// Return argument attributes with annotations removed.
-void GrandCentralVisitor::handlePorts(Operation *op) {
-
-  SmallVector<Attribute> newArgAttrs;
-  auto ports = cast<FModuleLike>(op).getPorts();
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    auto port = ports[i];
-    handleRefLike(op, port.annotations, port.type);
-    newArgAttrs.push_back(port.annotations.getArrayAttr());
-  }
-
-  op->setAttr("portAnnotations", ArrayAttr::get(op->getContext(), newArgAttrs));
-}
+/// Return the module that is associated with this value.
+///
+/// TODO: Refactor this to not use `getReferencedModule`.
+FModuleLike GrandCentralPass::getEnclosingModule(Value value) {
+  if (auto blockArg = value.dyn_cast<BlockArgument>())
+    return cast<FModuleOp>(blockArg.getOwner()->getParentOp());
 
-void GrandCentralVisitor::visitDecl(InstanceOp op) {
-
-  // If this instance's underlying module has a "companion" annotation, then
-  // move this onto the actual instance op.
-  AnnotationSet annotations(op.getReferencedModule());
-  if (auto anno = annotations.getAnnotation(
-          "sifive.enterprise.grandcentral.ViewAnnotation")) {
-    auto tpe = anno.getAs<StringAttr>("type");
-    if (!tpe) {
-      op.getReferencedModule()->emitOpError(
-          "contains a ViewAnnotation that does not contain a \"type\" field");
-      failed = true;
-      return;
-    }
-    if (tpe.getValue() == "companion")
-      op->setAttr("lowerToBind", BoolAttr::get(op.getContext(), true));
-  }
+  auto *op = value.getDefiningOp();
+  if (InstanceOp instance = dyn_cast<InstanceOp>(op))
+    return instance.getReferencedModule();
```

Comment:
I like the TODO suggestion of building a `SymbolTable` of the design up front. There's some precedence with the `CircuitNamespace` and `InstancePaths` :grinning: